File: GEMA_andorid/GEMAINVENTORY/app/src/main/java/com/example/gemainventory/ui/settings/AddSupplierFragment.java
package com.example.gemainventory.ui.settings;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;
import com.example.gemainventory.R;
import com.example.gemainventory.api.RetrofitClient;
import com.example.gemainventory.model.ProveedorDto;
import com.google.android.material.textfield.TextInputEditText;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class AddSupplierFragment extends Fragment {

    private TextInputEditText etName, etContact, etAddress;
    private Button btnSave;
    private String supplierId = null; // null = New

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_add_supplier, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        etName = view.findViewById(R.id.et_supplier_name);
        etContact = view.findViewById(R.id.et_supplier_contact);
        etAddress = view.findViewById(R.id.et_supplier_address);
        btnSave = view.findViewById(R.id.btn_save_supplier);

        view.findViewById(R.id.btn_back).setOnClickListener(v -> NavHostFragment.findNavController(this).popBackStack());

        // EDIT MODE
        if (getArguments() != null && getArguments().containsKey("supplier_id")) {
            supplierId = getArguments().getString("supplier_id");
            etName.setText(getArguments().getString("supplier_name"));
            etAddress.setText(getArguments().getString("supplier_address"));
            etContact.setText(getArguments().getString("supplier_contact"));

            ((android.widget.TextView) view.findViewById(R.id.tv_header_title)).setText("Editar Proveedor");
            btnSave.setText("Actualizar Proveedor");
        }

        btnSave.setOnClickListener(v -> guardarProveedor());
    }

    private void guardarProveedor() {
        String nombre = etName.getText().toString().trim();
        String contacto = etContact.getText().toString().trim();
        String direccion = etAddress.getText().toString().trim();

        if (nombre.isEmpty()) {
            etName.setError("Requerido");
            return;
        }

        SharedPreferences prefs = requireActivity().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE);
        String userId = prefs.getString("user_id", null);

        if (userId == null) {
            Toast.makeText(getContext(), "Error de sesión", Toast.LENGTH_SHORT).show();
            return;
        }

        btnSave.setEnabled(false);
        btnSave.setText("Guardando...");

        if (supplierId == null) {
            // CREATE
            ProveedorDto nuevo = new ProveedorDto(null, nombre, contacto, direccion);
            Call<ProveedorDto> call = RetrofitClient.INSTANCE.getInstance().crearProveedor(nuevo, userId);
            call.enqueue(new Callback<ProveedorDto>() {
                @Override
                public void onResponse(Call<ProveedorDto> call, Response<ProveedorDto> response) {
                    if (response.isSuccessful()) {
                        Toast.makeText(getContext(), "Proveedor guardado", Toast.LENGTH_SHORT).show();
                        NavHostFragment.findNavController(AddSupplierFragment.this).popBackStack();
                    } else {
                        btnSave.setEnabled(true);
                        btnSave.setText("Guardar Proveedor");
                        Toast.makeText(getContext(), "Error: " + response.code(), Toast.LENGTH_SHORT).show();
                    }
                }
                @Override
                public void onFailure(Call<ProveedorDto> call, Throwable t) {
                    btnSave.setEnabled(true);
                    btnSave.setText("Guardar Proveedor");
                    Toast.makeText(getContext(), "Error de conexión", Toast.LENGTH_SHORT).show();
                }
            });
        } else {
            // UPDATE
            ProveedorDto edit = new ProveedorDto(supplierId, nombre, contacto, direccion);
            Call<ProveedorDto> call = RetrofitClient.INSTANCE.getInstance().actualizarProveedor(supplierId, edit);
            call.enqueue(new Callback<ProveedorDto>() {
                @Override
                public void onResponse(Call<ProveedorDto> call, Response<ProveedorDto> response) {
                    if (response.isSuccessful()) {
                        Toast.makeText(getContext(), "Proveedor actualizado", Toast.LENGTH_SHORT).show();
                        NavHostFragment.findNavController(AddSupplierFragment.this).popBackStack();
                    } else {
                        btnSave.setEnabled(true);
                        btnSave.setText("Actualizar Proveedor");
                        Toast.makeText(getContext(), "Error al actualizar", Toast.LENGTH_SHORT).show();
                    }
                }
                @Override
                public void onFailure(Call<ProveedorDto> call, Throwable t) {
                    btnSave.setEnabled(true);
                    btnSave.setText("Actualizar Proveedor");
                    Toast.makeText(getContext(), "Error de conexión", Toast.LENGTH_SHORT).show();
                }
            });
        }
    }
}

File: GEMA_andorid/GEMAINVENTORY/app/src/main/java/com/example/gemainventory/ui/settings/AddSupplierFragment.kt
package com.example.gemainventory.ui.settings

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Business
import androidx.compose.material.icons.filled.ContactPhone
import androidx.compose.material.icons.filled.Home
import androidx.compose.runtime.*
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.example.gemainventory.api.RetrofitClient
import com.example.gemainventory.model.ProveedorDto
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class AddSupplierFragment : Fragment() {

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val prefs = requireContext().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE)
        val isDarkMode = prefs.getBoolean("DarkMode", false)
        val userId = prefs.getString("user_id", null)

        val args = arguments
        val supplierId = args?.getString("supplier_id")
        val initialName = args?.getString("supplier_name") ?: ""
        val initialAddress = args?.getString("supplier_address") ?: ""
        val initialContact = args?.getString("supplier_contact") ?: ""

        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                var name by remember { mutableStateOf(initialName) }
                var address by remember { mutableStateOf(initialAddress) }
                var contact by remember { mutableStateOf(initialContact) }
                var isLoading by remember { mutableStateOf(false) }

                val fields = listOf(
                    FormFieldData(
                        value = name,
                        onValueChange = { name = it },
                        label = "Nombre del Proveedor",
                        icon = Icons.Default.Business
                    ),
                    FormFieldData(
                        value = address,
                        onValueChange = { address = it },
                        label = "Dirección",
                        icon = Icons.Default.Home
                    ),
                    FormFieldData(
                        value = contact,
                        onValueChange = { contact = it },
                        label = "Contacto (Email o Teléfono)",
                        icon = Icons.Default.ContactPhone
                    )
                )

                GenericAddFormScreen(
                    title = if (supplierId == null) "Nuevo Proveedor" else "Editar Proveedor",
                    darkTheme = isDarkMode,
                    fields = fields,
                    buttonText = if (supplierId == null) "Guardar Proveedor" else "Actualizar Proveedor",
                    isLoading = isLoading,
                    onBackClick = { findNavController().popBackStack() },
                    onSaveClick = {
                        if (name.isBlank()) {
                            Toast.makeText(context, "El nombre es requerido", Toast.LENGTH_SHORT).show()
                        } else {
                            isLoading = true
                            saveSupplier(supplierId, name, contact, address, userId) { success ->
                                isLoading = false
                                if (success) findNavController().popBackStack()
                            }
                        }
                    }
                )
            }
        }
    }

    private fun saveSupplier(id: String?, name: String, contact: String, address: String, userId: String?, onComplete: (Boolean) -> Unit) {
        val dto = ProveedorDto(id, name, contact, address)
        val call = if (id == null) {
            RetrofitClient.instance.crearProveedor(dto, userId ?: "")
        } else {
            RetrofitClient.instance.actualizarProveedor(id, dto)
        }

        call.enqueue(object : Callback<ProveedorDto> {
            override fun onResponse(call: Call<ProveedorDto>, response: Response<ProveedorDto>) {
                if (response.isSuccessful) {
                    Toast.makeText(context, "Proveedor guardado", Toast.LENGTH_SHORT).show()
                    onComplete(true)
                } else {
                    Toast.makeText(context, "Error: ${response.code()}", Toast.LENGTH_SHORT).show()
                    onComplete(false)
                }
            }

            override fun onFailure(call: Call<ProveedorDto>, t: Throwable) {
                Toast.makeText(context, "Error de red", Toast.LENGTH_SHORT).show()
                onComplete(false)
            }
        })
    }
}
